From 3e055e198a35bdc433938848a5ab729ea03db1a2 Mon Sep 17 00:00:00 2001
From: Bin Han <bin.b.han@intel.com>
Date: Mon, 7 Dec 2015 15:40:53 +0800
Subject: [PATCH 09/13] ov8858: support V4L2_CID_LINK_FREQ for BTNS

V4L2_CID_LINK_FREQ is used to report MIPI freq to ISP driver
which helps ISP to configure MIPI receiver properly.

BTNS OV8858 has 360MHz MIPI freq,  thus report the value
through this control code.

Change-Id: I15aaa80e79c235637b1c6668442aebeb896fc661
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-13009
Original-Tracked-On: #H1504112909
Signed-off-by: Bin Han <bin.b.han@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/88504
Validation-Android: Wei Tao <wei.tao@intel.com>
Approver: Tian Shu Qiu <tian.shu.qiu@intel.com>
Build: Dessert M BXT RVP (SC)
Build: Dessert M CHT FFD (SC)
Build: Dessert M COHO (SC)
Validation-Android: Dessert M CHT FFD (SC)
Integrator: WeiX Meng <weix.meng@intel.com>
Signed-off-by: Yang Fan F <fan.f.yang@intel.com>
Reviewed-on: https://android.intel.com:443/450390
---
 .../camera/drivers/media/i2c/ov8858.c                | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c b/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
index f9748d4..c34c727 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
@@ -1933,6 +1933,11 @@ static int ov8858_g_ctrl(struct v4l2_ctrl *ctrl)
 	case V4L2_CID_EXPOSURE_ABSOLUTE:
 		ctrl->val = dev->exposure;
 		break;
+#ifdef CONFIG_PLATFORM_BTNS
+	case V4L2_CID_LINK_FREQ:
+		ctrl->val = 360000000;
+		break;
+#endif
 	default:
 		dev_warn(&client->dev,
 			 "%s: Error: Invalid ctrl: 0x%X\n", __func__, ctrl->id);
@@ -2253,7 +2258,20 @@ static const struct v4l2_ctrl_config ctrls[] = {
 		.min = V4L2_EXPOSURE_AUTO,
 		.max = V4L2_EXPOSURE_APERTURE_PRIORITY,
 		.step = 1,
-	}
+	},
+#ifdef CONFIG_PLATFORM_BTNS
+	{
+		.ops = &ctrl_ops,
+		.id = V4L2_CID_LINK_FREQ,
+		.name = "Link Frequency",
+		.type = V4L2_CTRL_TYPE_INTEGER,
+		.min = 1,
+		.max = 1500000000,
+		.step = 1,
+		.def = 360000000,
+		.flags = V4L2_CTRL_FLAG_VOLATILE | V4L2_CTRL_FLAG_READ_ONLY,
+	},
+#endif
 };
 
 static int ov8858_probe(struct i2c_client *client,
-- 
1.9.1

